bloom/infra/spire_service.py

```diff
@@ -0,0 +1,155 @@
+                    9249568
+                    8901913
+                ]
+            ) {{
+                pageInfo {{
+                    hasNextPage
+                    endCursor
+                }}
+                nodes {{
+                    id
+                    updateTimestamp
+                    staticData {{
+                        aisClass
+                        flag
+                        name
+                        callsign
+                        timestamp
+                        updateTimestamp
+                        shipType
+                        shipSubType
+                        mmsi
+                        imo
+                        callsign
+                        dimensions {{
+                            a
+                            b
+                            c
+                            d
+                            width
+                            length
+                        }}
+                    }}
+                    lastPositionUpdate {{
+                        accuracy
+                        collectionType
+                        course
+                        heading
+                        latitude
+                        longitude
+                        maneuver
+                        navigationalStatus
+                        rot
+                        speed
+                        timestamp
+                        updateTimestamp
+                    }}
+                    currentVoyage {{
+                        destination
+                        draught
+                        eta
+                        timestamp
+                        updateTimestamp
+                    }}
+                }}
+            }}
+        }}
+        """
+    
+    def get_raw_vessels(self, vessels: list[Vessel]):
+        query_string = self.create_query_string(vessels)
+        client = self.create_client()
+        paging = Paging()
+        hasNextPage: bool = False
+        raw_vessels = []
+        
+        while True:
+            try: 
+                response, hasNextPage = paging.page_and_get_response(client, query_string)
+                if response:
+                    raw_vessels += response.get('vessels', {}).get('nodes', [])
+                    if not hasNextPage:
+                        break                
+            except BaseException as e:
+                logger.error(e)
+                raise
+
+        return raw_vessels
+
+
+        
+    
+    def get_vessels(self, vessels: list[Vessel]):
+        raw_vessels = self.get_raw_vessels(vessels)
+        formatted_vessels = self.format_raw_vessels(raw_vessels)
+
+        return formatted_vessels
+
+    def format_raw_vessel(self, raw_vessel) -> Vessel:
+        return raw_vessel
+
+    def format_raw_vessels(self, raw_vessels) -> list[Vessel]:
+        vessels = []
+
+        for raw_vessel in raw_vessels:
+            vessels.append(self.format_raw_vessel(raw_vessel))
```

Comment:
This could also be `[self.format_raw_vessel(raw_vessel) for raw_vessel in raw_vessels]` am I right? What's the best option?

Reply:
https://towardsdatascience.com/list-comprehensions-vs-for-loops-it-is-not-what-you-think-34071d4d8207